pkg/velerorunner/task.go

```diff
@@ -0,0 +1,382 @@
+	Restore         *velero.Restore
+	ClientCache     map[*migapi.MigCluster]k8sclient.Client
+}
+
+// Reconcile() Example:
+//
+// task := velerorunner.Task{
+//     Log: log,
+//     Client: r,
+//     Owner: migration,
+//     PlanResources: plan.GetPlanResources(),
+// }
+//
+// completed, err := task.Run()
+//
+
+// Run the task.
+// Return `true` when run to completion.
+func (t *Task) Run() (bool, error) {
+	// Backup
+	err := t.ensureBackup()
+	if err != nil {
+		return false, err
+	}
+	if t.Backup.Status.Phase != velero.BackupPhaseCompleted {
+		t.Log.Info(
+			"Waiting for backup to complete.",
+			"owner",
+			t.Owner.GetName(),
+			"backup",
+			t.Backup.Name)
+		return false, nil
+	}
+	t.Log.Info(
+		"Backup has completed.",
+		"owner",
+		t.Owner.GetName(),
+		"backup",
+		t.Backup.Name)
+
+	backup, err := t.getReplicaatedBackup()
+	if err != nil {
+		return false, err
+	}
+	if backup == nil {
+		t.Log.Info(
+			"Waiting for backup to be replicated to the destination.",
+			"owner",
+			t.Owner.GetName(),
+			"backup",
+			t.Backup.Name)
+		return false, nil
+	}
+	// Restore
+	err = t.ensureRestore()
+	if err != nil {
+		return false, err
+	}
+	if t.Restore.Status.Phase != velero.RestorePhaseCompleted {
+		t.Log.Info(
+			"Waiting for restore to complete.",
+			"owner",
+			t.Owner.GetName(),
+			"restore",
+			t.Restore.Name)
+		return false, nil
+	}
+	t.Log.Info(
+		"Restore has completed.",
+		"owner",
+		t.Owner.GetName(),
+		"restore",
+		t.Restore.Name)
+
+	return true, nil
+}
+
+// Ensure the backup on the source cluster has been created
+// and has the proper settings.
+func (t *Task) ensureBackup() error {
+	newBackup, err := t.buildBackup()
+	if err != nil {
+		return err
+	}
+	foundBackup, err := t.getBackup()
+	if err != nil {
+		return err
+	}
+	if foundBackup == nil {
+		t.Backup = newBackup
+		client, err := t.getSourceClient()
+		if err != nil {
+			return err
+		}
+		err = client.Create(context.TODO(), newBackup)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	t.Backup = foundBackup
+	if !t.equalsBackup(newBackup, foundBackup) {
+		client, err := t.getSourceClient()
+		if err != nil {
+			return err
+		}
+		t.updateBackup(foundBackup)
+		err = client.Update(context.TODO(), foundBackup)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get whether the two Backups are equal.
+func (t *Task) equalsBackup(a, b *velero.Backup) bool {
+	match := a.Spec.StorageLocation == b.Spec.StorageLocation &&
+		reflect.DeepEqual(a.Spec.VolumeSnapshotLocations, b.Spec.VolumeSnapshotLocations) &&
+		reflect.DeepEqual(a.Spec.IncludedNamespaces, b.Spec.IncludedNamespaces) &&
+		reflect.DeepEqual(a.Spec.IncludedResources, b.Spec.IncludedResources) &&
+		a.Spec.TTL == b.Spec.TTL
+	return match
+}
+
+// Get an existing Backup on the source cluster.
+func (t Task) getBackup() (*velero.Backup, error) {
+	client, err := t.getSourceClient()
+	if err != nil {
+		return nil, err
+	}
+	list := velero.BackupList{}
```

Comment:
Does this pull every backup in its entirety off the source cluster? If so, once there get to be a lot of backups, could be a pretty heavy operation. Something to keep an eye on.

Reply:
Perhaps this is just pulling the names of Backups off, rather than all of the contents. Pulling just the names off probably isn't a big deal.

Reply:
No, just ones matching the label selection.  should never be more than 1.